dcos/http/client/client.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2016 Mesosphere, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"os"
+
+	"github.com/dcos/dcos-go/jwt/transport"
+)
+
+// loadCAPool will load a valid x509 cert.
+func loadCAPool(path string) (*x509.CertPool, error) {
+	caPool := x509.NewCertPool()
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	b, err := ioutil.ReadAll(f)
+	if err != nil {
+		return nil, err
+	}
+
+	if !caPool.AppendCertsFromPEM(b) {
+		return nil, errors.New("CACertFile parsing failed")
+	}
+
+	return caPool, nil
+}
+
+// getTransport will return transport for http.Client
+func getTransport(caCertificatePath string) (*http.Transport, error) {
+	tr := &http.Transport{}
+	// if user provided CA cert we must use it, otherwise use InsecureSkipVerify: true for all HTTPS requests.
+	if caCertificatePath != "" {
+		caPool, err := loadCAPool(caCertificatePath)
+		if err != nil {
+			return tr, err
+		}
+
+		tr.TLSClientConfig = &tls.Config{
+			RootCAs: caPool,
+		}
+	} else {
+		tr.TLSClientConfig = &tls.Config{
+			InsecureSkipVerify: true,
+		}
+	}
+	return tr, nil
+}
+
+// NewDCOSClient returns a client with CAPool when passed a CA certificate path and
+// a Authorization header when passed a IAM configuration path. If the path to CA
+// certificate or IAM config is an empty string then the corrosponding client
+// configurations which they set are ignored.
+func NewDCOSClient(caCertificatePath string, iamConfigPath string) (*http.Client, error) {
```

Comment:
you could split the baby:
```go
// NewTransport creates and returns an HTTP configured with the given certs and (optional) IAM configuration
func NewTransport(caCertificatePath string, iamConfigPath string) (http.RoundTripper, error) {
  // most of the code from the existing NewDCOSClient (just don't create an http.Client)
  tr, err := newTransport(caCertificatePath)
  if err != nil {
	return nil, err
  }
  if len(iamConfigPath) != 0 {
    withIAM, err := transport.NewRoundTripper(
      tr,
      transport.OptionReadIAMConfig(iamConfigPath),
    )
    if err != nil {
      return nil, err
    }
    return withIAM
  }
  return tr
}

// NewClient creates and returns an HTTP client (based on http.DefaultClient), configured with an HTTP transport via NewTransport
func NewClient(caCertificatePath string, iamConfigPath string) (*http.Client, error) {
  tr, err := NewTransport(caCertificatePath, iamConfigPath)
  if err != nil {
    return nil, err
  }
  client := *http.DefaultClient
  client.Transport = tr
  return &client, nil
}
```

you might even consider refactoring `newTransport` to accept an optional (nil-able) `*http.Transport` param. if it's nil, then use `&http.Transport{}` like you do now. once that's in place, you could add a similar optional (nil-able) param to `NewTransport` so that a caller could provide the base transport layer.

(notice s/NewDCOSClient/NewClient/ to reduce stutter)